### PR TITLE
Feature/issue 1232 value of rec metaprogram

### PR DESCRIPTION
**Pull request by rtrangucci**
#### Summary:

Addresses #1232 

#### Intended Effect:

Adds a function value_of_rec that returns the base type of a variable. So named because 
it is the recursive version of value_of.  I followed the blueprint of value_of to create value_of_rec.

#### How to Verify:

$ ./runTests.py -j2 src/test/unit/math/functions/value_of_rec_test.cpp 
$ ./runTests.py -j2 src/test/unit/agrad/rev/functions/value_of_rec_test.cpp
$ ./runTests.py -j2 src/test/unit/agrad/fwd/functions/value_of_rec_test.cpp

$ ./runTests.py -j2 src/test/unit/math/matrix/value_of_rec_test.cpp   
$ ./runTests.py -j2 src/test/unit/agrad/rev/matrix/value_of_rec_test.cpp 
$ ./runTests.py -j2 src/test/unit/agrad/fwd/matrix/value_of_rec_test.cpp 

#### Side Effects:

None.

#### Documentation:

Added doxygen documentation for the functions, ripped mostly from value_of's doc.

#### Reviewer Suggestions: 

Anyone.

### CODE REVIEW

File: src/stan/agrad/fwd/matrix/value_of_rec.hpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+     * #include<stan/agrad/rev/functions/value_of_rec.hpp>
+     *
+     * @tparam T Scalar type in matrix
+     * @tparam R Rows of matrix
+     * @tparam C Columns of matrix
+     * @param[in] M Matrix to be converted
+     **/
+    template<typename T, int R,int C>
+    inline Eigen::Matrix<double,R,C> value_of_rec(const Eigen::Matrix<T,R,C> &M) {
+      int i,j;
```
Comment (bob-carpenter) [inline src/stan/agrad/fwd/matrix/value_of_rec.hpp]: These integer variables should be defined in the loops.  If you copied this from elsewhere, change where it was copied from, too, if you don't mind.

File: src/stan/agrad/rev/functions/value_of_rec.hpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+namespace stan {
+  namespace agrad {
+
+    /**
+     * Return the value of the specified variable.  
+     *
+     * @param v Variable.
+     * @return Value of variable.
+     */
+    inline double value_of_rec(const stan::agrad::var& v) {
```
Comment (bob-carpenter) [inline src/stan/agrad/rev/functions/value_of_rec.hpp]: When you're in `stan::agrad` namespace you don't need `stan::agrad` qualifier on `var`.

File: src/stan/agrad/rev/matrix/value_of_rec.hpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+     * a new matrix is returned with the value of the constituent vars.
+     *
+     * This is used as a convenience function for implementing varis of matrix operations.
+     * @tparam R Rows of matrix M
+     * @tparam C Columns of matrix M
+     * @param[in] M Matrix
+     **/
+    template<int R,int C>
+    inline Eigen::Matrix<double,R,C> value_of_rec(const Eigen::Matrix<var,R,C> &M) {
+      int i,j;
```
Comment (bob-carpenter) [inline src/stan/agrad/rev/matrix/value_of_rec.hpp]: Same fix as before.

File: src/stan/math/matrix/value_of_rec.hpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    /**
+     * Convert a matrix to a matrix of doubles.
+     * Math double version is the identity function 
+     *
+     * @tparam R Rows of matrix
+     * @tparam C Columns of matrix
+     * @param[in] M Matrix to be converted
+     **/
+    template<int R,int C>
+    inline const Eigen::Matrix<double,R,C> &value_of_rec(const Eigen::Matrix<double,R,C> &M) {
```
Comment (bob-carpenter) [inline src/stan/math/matrix/value_of_rec.hpp]: `>&` rather than `> &`

`template<` spacing

File: src/stan/agrad/rev/matrix/value_of.hpp [hunk 5]
```diff
@@ -23,12 +23,11 @@ namespace stan {
      *
      * This is used as a convenience function for implementing varis of matrix operations.
      **/
-    template<int R,int C>
-    inline Eigen::Matrix<double,R,C> value_of(const Eigen::Matrix<var,R,C> &M) {
-      int i,j;
+    template<int R, int C>
```
Comment (bob-carpenter) [inline src/stan/agrad/rev/matrix/value_of.hpp]: I still want a space between `template` and `<`.

File: src/stan/agrad/rev/matrix/value_of.hpp [hunk 6]
```diff
@@ -23,12 +23,11 @@ namespace stan {
...
+    inline Eigen::Matrix<double,R,C> value_of(const Eigen::Matrix<var,R,C>& M) {
```
Comment (bob-carpenter) [inline src/stan/agrad/rev/matrix/value_of.hpp]: This whole definition should go away  It's redundant given the template function in

`src/stan/math/matrix/value_of_rec.hpp`

Comment (bob-carpenter) [inline src/stan/agrad/rev/matrix/value_of.hpp]: The same thing should be done with value_of as with value_of_rec.  There's no need for a var-specific Matrix definition here.

**Comment by bob-carpenter** [comment]
Just to repeat what I said in a code line note, the same thing should be done to make the `value_of` function generic for matrices.  No need to define one for matrix of var and another for matrix of fvar.

File: src/stan/agrad/fwd/functions/value_of_rec.hpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
+#ifndef STAN__AGRAD__FWD__FUNCTIONS__VALUE_OF_REC_HPP
+#define STAN__AGRAD__FWD__FUNCTIONS__VALUE_OF_REC_HPP
+
+#include <stan/math/functions/value_of_rec.hpp>
```
Comment (bob-carpenter) [inline src/stan/agrad/fwd/functions/value_of_rec.hpp]: Need to include fvar for the header to compile.

File: src/stan/agrad/rev/functions/value_of_rec.hpp [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#ifndef STAN__AGRAD__REV__FUNCTIONS__VALUE_OF_REC_HPP
+#define STAN__AGRAD__REV__FUNCTIONS__VALUE_OF_REC_HPP
+
```
Comment (bob-carpenter) [inline src/stan/agrad/rev/functions/value_of_rec.hpp]: needs to include var

**Comment by syclik** [comment]
Jenkins, retest this please.

**Comment by syclik** [comment]
Jenkins, retest this please.

**Comment by syclik** [comment]
@rtrangucci, check this page to see what's going on:
http://d1m1s1b1.stat.columbia.edu:8080/job/Stan%20Pull%20Request%20-%20Tests%20-%20Unit/169/warnings17Result/new/

**Pull request merged**